### PR TITLE
fixes #6856 - API v2 - more efficient import puppetclasses for single environment

### PR DESCRIPTION
**Pull request by isratrade**

### CODE REVIEW

**Comment by dLobatog** [comment]
Works well, thanks. Could you add a test?

**Comment by isratrade** [comment]
@eLobato, added test.

File: test/functional/api/v2/smart_proxies_controller_test.rb [hunk 1]
```diff
@@ -137,6 +137,22 @@ def setup_import_classes
...
+  test "should import only one specified new environment" do
+    setup_import_classes
+    as_admin do
+      Host::Managed.update_all(:environment_id => nil)
+      Hostgroup.update_all(:environment_id => nil)
+      Puppetclass.destroy_all
+      Environment.destroy_all
+    end
+    assert_difference('Environment.count', 1) do
+      post :import_puppetclasses, {:id => smart_proxies(:puppetmaster).id, :environment_id => 'env1'}, set_session_user
```
Comment (dLobatog) [inline test/functional/api/v2/smart_proxies_controller_test.rb]: This test always passes with and without the patch. Environments are imported properly when the parameter is passed. What is not imported properly is puppetclasses. 

A suggestion on how to test this is to make `env1` return puppetclass `a` and make `env2` return puppetclass `b`, and then check that this query only adds puppetclass `a` to the list of our puppetclasses, and not `b`. Removing the environments for this test is not needed, only the puppetclasses.

In order to do this you'll need to change some expectations of ProxyAPI::Puppet, since `setup_import_classes` makes it import the same puppetclasses for all environments.

File: test/functional/api/v2/smart_proxies_controller_test.rb [hunk 2]
```diff
@@ -137,6 +136,30 @@ def setup_import_classes
...
+      HostgroupClass.destroy_all
+      Puppetclass.destroy_all
+      Environment.destroy_all
+    end
+
+    ProxyAPI::Puppet.any_instance.stubs(:environments).returns(["env1", "env2"])
+    classes_env1 = {'a' => Foreman::ImporterPuppetclass.new('name' => 'a')}
+    classes_env2 = {'b' => Foreman::ImporterPuppetclass.new('name' => 'b')}
+    ProxyAPI::Puppet.any_instance.stubs(:classes).with('env1').returns(classes_env1)
+    ProxyAPI::Puppet.any_instance.stubs(:classes).with('env2').returns(classes_env2)
```
Comment (isratrade) [inline test/functional/api/v2/smart_proxies_controller_test.rb]: @eLobato, updated the test in commit to squash. What hung me up is that even though I don't check for `classes.with('env2')`, I needed to declare it or there was an error.

Comment (isratrade) [inline test/functional/api/v2/smart_proxies_controller_test.rb]: @eLobato, note that I don't call `setup_import_classes` on this test. Whenever I changed this, it messed up other tests.

Comment (dLobatog) [inline test/functional/api/v2/smart_proxies_controller_test.rb]: Right.. there's something wrong I just found. You should remove the stub for classes_env2 as it shouldn't be called at all, but with the current code it is called. Let me comment in other places why this is happening.

File: test/functional/api/v2/smart_proxies_controller_test.rb [hunk 3]
```diff
@@ -137,6 +136,30 @@ def setup_import_classes
     assert_equal 2, response['environments_with_new_puppetclasses']
   end
 
+  test "should import puppetclasses for specified environment only" do
+    as_admin do
+      Host::Managed.update_all(:environment_id => nil)
```
Comment (dLobatog) [inline test/functional/api/v2/smart_proxies_controller_test.rb]: The whole `as_admin` block is unnecessary. No `env1` or `env2` puppetclasses exist at this point, only the fixtures. It made sense with the setup_import_classes, now it's just a leftover.

File: app/controllers/concerns/api/import_puppetclasses_common_controller.rb [hunk 4]
```diff
@@ -49,6 +49,7 @@ def import_puppetclasses
   def changed_environments
     begin
       opts      =  { :url => @smart_proxy.url }
+      opts.merge!(:env => @environment.name) if @environment
```
Comment (dLobatog) [inline app/controllers/concerns/api/import_puppetclasses_common_controller.rb]: `@environment` is not set at this point for your tests - which is OK and I'm glad your tests revealed this. 
The reason why it's not set is that the before_filter `find_optional_environment` will not find environments like `env1` or `env2` in your tests, simply because they don't exist yet. 

Instead, let's use 

``` ruby
if @environment.present?
  opts.merge!(:env => @environment.name)
else
  opts.merge!(:env => @env_id)
end
```

`@env_id` is always defined, and it can be an environment that exists or doesn't exist, it doesn't matter. The Proxy will import the classes and environment properly if it doesn't exist. And if it exists, it will use the handy `@environment` variable because `find_optional_environment` worked.

Summing up, `@env_id` will be used when users try to import classes from environments that are not in Foreman, and `@enviroment` will be used to import classes from environments existing in Foreman. From the user point of view, this is seamless and the use of these variables is 100% internal and hidden, they have the same API as before (even better, now they can import classes from non existing environments)

**Comment by isratrade** [comment]
@eLobato, thanks for the comments. I agree. I added a second test as well when no environment is specified.

File: test/functional/api/v2/smart_proxies_controller_test.rb [hunk 5]
```diff
@@ -137,6 +136,35 @@ def setup_import_classes
     assert_equal 2, response['environments_with_new_puppetclasses']
   end
 
+  test "should import puppetclasses for specified environment only" do
+    ProxyAPI::Puppet.any_instance.stubs(:environments).returns(["env1", "env2"])
+    classes_env1 = {'a' => Foreman::ImporterPuppetclass.new('name' => 'a')}
+    classes_env2 = {'b' => Foreman::ImporterPuppetclass.new('name' => 'b')}
```
Comment (dLobatog) [inline test/functional/api/v2/smart_proxies_controller_test.rb]: `classes_env2` is not used at all in this test. 

Aside from that, this looks very good, one last change before merging could be putting these two tests in the same `context` and adding all stubs in the before block. Does that sound OK to you?

Comment (isratrade) [inline test/functional/api/v2/smart_proxies_controller_test.rb]: `classes_env2` is not used, but remember the previous test failed without the stub since it was still being called. I left it in. Otherwise, the test seems meaningless as only `classes_env1` exists and therefore only `classes_env1` would appear whether an `environment_id` parameter was passed in the URL or not.  Does this make sense?

Comment (dLobatog) [inline test/functional/api/v2/smart_proxies_controller_test.rb]: That's right, so that's the point of putting stubs in a context between this test and the following. This test will get stubs for env2 and it will only import env1. The test is not meaningless as it is though, because without the fix, `ProxyAPI::Puppet.classes('env2')` would get called and make the test fail.

File: test/functional/api/v2/smart_proxies_controller_test.rb [hunk 6]
```diff
@@ -250,4 +249,32 @@ def setup_import_classes
     response = ActiveSupport::JSON.decode(@response.body)
   end
 
+  context 'import puppetclasses' do
+    setup do
+      ProxyAPI::Puppet.any_instance.stubs(:environments).returns(["env1", "env2"])
+      classes_env1 = {'a' => Foreman::ImporterPuppetclass.new('name' => 'a')}
+      classes_env2 = {'b' => Foreman::ImporterPuppetclass.new('name' => 'b')}
+      ProxyAPI::Puppet.any_instance.stubs(:classes).with('env1').returns(classes_env1)
+      ProxyAPI::Puppet.any_instance.stubs(:classes).with('env2').returns(classes_env2)
```
Comment (isratrade) [inline test/functional/api/v2/smart_proxies_controller_test.rb:258]: @eLobato, added context block, but a side effect is now both tests have stubs for returning `classes_env1` and `classes_env2`

Comment (dLobatog) [inline test/functional/api/v2/smart_proxies_controller_test.rb:258]: That's fine, the first test will just fail the `refuse_includes` now, without the stub it'll raise an exception (both are failures). :smile: merge in 3...2..1

**Comment by dLobatog** [comment]
[test] , I'm not sure why it failed on mysql 1.8.7.

File: test/functional/api/v2/smart_proxies_controller_test.rb (hunk 6 above)
Comment (isratrade) [inline test/functional/api/v2/smart_proxies_controller_test.rb:258]: @eLobato, not following, you write that the both tests will fail?? Both of the tests I added passed, but there is an obscure failure for 1.8.7/mysql http://ci.theforeman.org/job/test_develop_pr_core/2532/database=mysql,puppet=3.0,ruby=1.8.7/
I'll squash and rebase.

Comment (dLobatog) [inline test/functional/api/v2/smart_proxies_controller_test.rb:258]: Sorry I didn't explain myself clearly, what I meant was that the first test would fail with or without the stubs **IF** future updates to the code break this behavior.

**Comment by dLobatog** [comment]
Merged as https://github.com/theforeman/foreman/commit/4c59b8daccf3534fa89a31076b3e2ca4ae7fdcee , thanks @isratrade !
